synapseutils/copy.py

```diff
@@ -2,55 +2,124 @@
-    
-    :param contentTypes:            List of content types (Can change a filetype of a filehandle).
 
-    :param fileNames:               List of filenames (Can change a filename of a filehandle).
-    
+    :param newContentTypes:         (Optional) List of content types (Can change a filetype of a filehandle).
+
+    :param newFileNames:            (Optional) List of filenames (Can change a filename of a filehandle).
+
     :return:                        List of batch filehandle copy results, can include failureCodes: UNAUTHORIZED and
                                     NOT_FOUND
     """
-    if (len(fileHandles) != len(associateObjectTypes) or len(fileHandles) != len(associateObjectIds)
-            or len(fileHandles) != len(contentTypes) or len(fileHandles) != len(fileNames)):
-        raise ValueError("Length of fileHandles, associateObjectTypes, and associateObjectIds must be the same")
-    fileHandles = [synapseclient.core.utils.id_of(handle) for handle in fileHandles]
-    copyFileHandleRequest = {"copyRequests": []}
-    for filehandleId, contentType, fileName, associateObjectType, associateObjectId \
-            in zip(fileHandles, contentTypes, fileNames, associateObjectTypes, associateObjectIds):
-        copyFileHandleRequest['copyRequests'].append({"newContentType": contentType,
-                                                      "newFileName": fileName,
-                                                      "originalFile": {"associateObjectType": associateObjectType,
-                                                                       "fileHandleId": filehandleId,
-                                                                       "associateObjectId": associateObjectId}})
-    copiedFileHandles = syn.restPOST('/filehandles/copy', body=json.dumps(copyFileHandleRequest),
-                                     endpoint=syn.fileHandleEndpoint)
-    _copy_cached_file_handles(syn.cache, copiedFileHandles)
-    return copiedFileHandles
+
+    # Check if length of all inputs are equal
+    if not (len(fileHandles) == len(associateObjectTypes) == len(associateObjectIds)
+            and (newContentTypes is None or len(newContentTypes) == len(associateObjectIds))
+            and (newFileNames is None or len(newFileNames) == len(associateObjectIds))):
+        raise ValueError("Length of all input arguments must be the same")
+
+    # If no optional params passed, assign to empty list
+    if newContentTypes is None:
+        newContentTypes = []
+    if newFileNames is None:
+        newFileNames = []
+
+    # Remove this line if we change API to only take fileHandleIds and not Objects
+    file_handle_ids = [synapseclient.core.utils.id_of(handle) for handle in fileHandles]
+
+    # add division logic for POST call here
+    master_copy_results_list = []  # list which holds all results from POST call
+    num_batches = math.ceil(len(fileHandles) / MAX_FILE_HANDLE_PER_COPY_REQUEST)
+    for i in range(num_batches):
```

Comment:
please find out if there is an exist utils to read next batch or refactor reading next batch into its own function